lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,244 @@
+require 'securerandom'
+require 'open3'
+require 'ffi'
+
+module PuppetX
+  module PowerShell
+    class PowerShellManager
+      extend Puppet::Util::Windows::String
+      extend FFI::Library
+
+      @@instances = {}
+
+      def self.instance(cmd)
+        @@instances[:cmd] ||= PowerShellManager.new(cmd)
+      end
+
+      def initialize(cmd)
+        @stdin, @stdout, @ps_process = Open3.popen2(cmd)
+
+        Puppet.debug "#{Time.now} #{cmd} is running as pid: #{@ps_process[:pid]}"
+
+        at_exit { exit }
+      end
+
+      def execute(powershell_code, timeout_ms = 300 * 1000)
+        output_ready_event_name =  "Global\\#{SecureRandom.uuid}"
+        output_ready_event = self.class.create_event(output_ready_event_name)
+
+        # always need a trailing newline to ensure PowerShell parses code
+        code = <<-CODE
+        $event = [Threading.EventWaitHandle]::OpenExisting("#{output_ready_event_name}")
+        if ($runspace -eq $null)
+        {
+          $runspace = [RunspaceFactory]::CreateRunspace()
+          $runspace.Open()
+        }
+
+        $powershell_code = @'
+#{powershell_code}
+'@
+        $ps = $null
+
+        try
+        {
+          # http://learn-powershell.net/2012/05/13/using-background-runspaces-instead-of-psjobs-for-better-performance/
+          $ps = [powershell]::create()
+          $ps.Runspace = $runspace
+          [Void]$ps.AddScript($powershell_code)
+
+          $asyncResult = $ps.BeginInvoke()
+
+          if (!$asyncResult.AsyncWaitHandle.WaitOne(#{timeout_ms}, $false))
+          {
+            throw "Catastrophic failure: PowerShell DSC resource timeout (#{timeout_ms} ms) exceeded while executing"
+          }
+
+          $output = $ps.EndInvoke($asyncResult)
+
+          if($ps.HadErrors){
+            [string[]]$msg = ""
+            $errors = $ps.Streams.Error.ReadAll()
+            foreach ($record in $errors) {
+              $msg += $record.InvocationInfo.MyCommand.Name + " : " + $record.Exception.Message
+              $msg += $record.InvocationInfo.PositionMessage
+              $msg += "+ CategoryInfo: " + $record.CategoryInfo
+              $msg += "+ FullyQualifiedErrorId: " + $record.FullyQualifiedErrorId
+            }
+            return @{
+              output       = $null
+              exitcode     = 1  # TODO: this is lying
+              errormessage = [string]::join("\n", $msg)
+            } | ConvertTo-Json -Compress
+          }
+
+          return @{
+            output       = $output
+            exitcode     = 0 # TODO: this is lying
+            errormessage = $null
+          } | ConvertTo-Json -Compress
+
+        }
+        catch
+        {
+          try
+          {
+            if ($runspace) { $runspace.Dispose() }
+          }
+          finally
+          {
+            $runspace = $null
+          }
+          @{
+            output       = $null
+            exitcode     = 1  # TODO: this is lying
```

Comment:
We are technically lying here. Since we keep PS open, there isn't any exit code or return code. Only binaries return exit codes in PS, cmdlets and the like do not.